src/routes/safe-apps/__tests__/get-safe-apps.e2e-spec.ts

```diff
@@ -0,0 +1,86 @@
+import * as request from 'supertest';
+import { RedisClientType } from 'redis';
+import { INestApplication } from '@nestjs/common';
+import { Test } from '@nestjs/testing';
+import { AppModule } from '../../../app.module';
+import { readFileSync } from 'fs';
+import { SafeApp } from '../entities/safe-app.entity';
+import { redisClientFactory } from '../../../../test/common/cache/redis-client.factory';
+
+describe('Get Safe Apps e2e test', () => {
+  let app: INestApplication;
+  let redisClient: RedisClientType;
+  const chainId = '5'; // Görli testnet
+
+  beforeAll(async () => {
+    const moduleRef = await Test.createTestingModule({
+      imports: [AppModule],
+    }).compile();
+
+    app = moduleRef.createNestApplication();
+    await app.init();
+    redisClient = await redisClientFactory();
+  });
+
+  beforeEach(async () => {
+    await redisClient.flushAll();
+  });
+
+  it('GET /chains/<chainId>/safe-apps', async () => {
+    const safeAppsCacheKey = `${chainId}_safe_apps`;
+    const safeAppsCacheField = 'undefined_undefined';
+    const expectedResponse: SafeApp[] = JSON.parse(
+      readFileSync(
+        'src/routes/safe-apps/__tests__/resources/get-all-safe-apps-expected-response.json',
+        {
+          encoding: 'utf-8',
+        },
+      ),
+    );
+
+    await request(app.getHttpServer())
+      .get(`/chains/${chainId}/safe-apps`)
+      .expect(200)
+      .then(({ body }) => {
+        expect(body).toEqual(expectedResponse);
+      });
+
+    const cacheContent = await redisClient.hGet(
+      safeAppsCacheKey,
+      safeAppsCacheField,
+    );
+    expect(cacheContent).toBeDefined();
+  });
+
+  it('GET /chains/<chainId>/safe-apps?url=https://app.ens.domains', async () => {
+    const safeAppsCacheKey = `${chainId}_safe_apps`;
+    const safeAppsCacheField = 'undefined_https://app.ens.domains';
+    const expectedResponse: SafeApp[] = JSON.parse(
+      readFileSync(
+        'src/routes/safe-apps/__tests__/resources/get-safe-apps-filtered-by-url-expected-response.json',
+        {
+          encoding: 'utf-8',
+        },
+      ),
+    );
+
+    await request(app.getHttpServer())
+      .get(`/chains/${chainId}/safe-apps/?url=https://app.ens.domains`)
+      .expect(200)
+      .then(({ body }) => {
+        expect(body).toEqual(expectedResponse);
+      });
+
+    const cacheContent = await redisClient.hGet(
+      safeAppsCacheKey,
+      safeAppsCacheField,
+    );
+    expect(cacheContent).toBeDefined();
+  });
+
+  afterAll(async () => {
+    await app.close();
+    await redisClient.flushAll();
+    await redisClient.quit();
+  });
+});
```

Comment:
Safe Apps can change frequently which makes this test unreliable.

I would not consider this endpoint for end to end tests unless it's with an app that we own and provide long term support.

Reply:
Yes, that's something I also considered, I thought that maybe staging shouldn't change so much but you're right, having a fixed JSON to match against changing data it isn't reliable. What do you think about just filtering by Transaction Builder (an app we own as you said)? Another option could be just validate the response structure (an array of objects with a id, url, name...)

Do you see any of this options feasible? Otherwise I could remove the test completely 🤔

Reply:
I would say that the test itself is valuable and one option is mocking the networking part so you cover the route.

I will leave this up to you. I think testing the Transaction Builder should be fine but I'm not 100% confident that it will work long term as the staging environment can be used for testing different configurations.

One thing is blockchain data which is the same between our environments. But metadata related with the blockchain (contract names, chain data, safe apps) are more subject to changes.

Reply:
Changed in [d9d1356](https://github.com/5afe/safe-client-gateway-nest/commit/d9d1356e98c2b05721a2e309e27194d2c784757f), I removed the JSON fixed payloads I'm just verifying the responses contain the expected data structure, and also, in the case of the Transaction Builder Safe App, the response contains just 1 element with the expected name and url (that shouldn't change a lot I assume). Does it sound good?

Reply:
The test `GET /chains/<chainId>/safe-apps?url=${transactionBuilderUrl}` fails for me. Which configuration are you using to test it? I believe this might be related with the config service url.

The configuration should probably be set by the test instead of relying on the local environment.

Reply:
I added default values for e2e/ci tests execution in [a36e858](https://github.com/5afe/safe-client-gateway-nest/pull/169/commits/a36e858bd88b6d1cf91311cfdf364fa59e7e711f)